Sources/FoundationNetworking/URLSession/URLSessionTask.swift

```diff
@@ -364,11 +364,20 @@ open class URLSessionTask : NSObject, NSCopying {
      */
     open func cancel() {
         workQueue.sync {
-            guard self.state == .running || self.state == .suspended else { return }
-            self.state = .canceling
+            let (canceled, request) = self.syncQ.sync { () -> (Bool, URLRequest?) in
+                guard self._state == .running || self._state == .suspended else { return (true, nil)}
+                _state = .canceling
+                return (false, _currentRequest ?? originalRequest)
```

Comment:
@spevans: Should the URL included in the error just be the original URL? I'm not certain of the expectation in the case of redirections, so I wonder if this is the correct data.

Reply:
If you think `_currentRequest ?? originalRequest` is correct, then this PR should be ready.

Reply:
Looking at some of the other tests with errors in redirection it looks like you should just use `originalRequest?.url` and dont bother with `_currentRequest`. Im pretty sure that `originalRequest` is never `nil` at the point the task can be cancelled so there is no need for the `??` fallback

Reply:
Great. I'll change it to just use `originalRequest`.